### PR TITLE
Refactoring training models

### PR DESCRIPTION
**Pull request by lizgzil**
https://trello.com/c/ybAUoW2W/1199-tech-model-refactor-train-multiple-models

Part of https://github.com/wellcometrust/nutrition-labels/issues/62

The main changes are in `grant_tagger.py`.

Previously we ran `grant_tagger.py` with many different arguments to train all the different model vectoriser + classifier combinations. Thus we had to run it 12 times with different arguments.

In this PR:
- `grant_tagger.py` will now train several models with details given in a config file, so only need to run once.
- The GrantTagger class isn't that different - just parameterising certain values, including probabilities in outputs, outputting train/test information (which data points were in test or train and the predictions for each model)
- Tests added for GrantTagger
- Updated README
- Docstrings
- No longer 'bert_type' parameter - just finds this from vectorizer. Before you had to do `vectorizer_type='bert'` and `bert_type='scibert'` to use scibert, but now its just `vectorizer_type='scibert'`.

### CODE REVIEW

**Comment by lizgzil** [comment]
[Tech model refactor - train multiple models](https://trello.com/c/ybAUoW2W/1199-tech-model-refactor-train-multiple-models)

File: README.md [hunk 1]
```diff
@@ -60,13 +60,19 @@ with config files from '2020.08.07.ini', '2021.01.26.ini', '2021.02.21.ini' or 2
...
+You can train and save a model, or several models, to classify grants as being to do with tech or not (see definitions for this in [Finding_Tech_Grants.md](docs/Finding_Tech_Grants.md)) by creating a config file which contains various training arguments for the experiments. The main arguments in this config file are:
 
+- `training_data_file`, e.g. `data/processed/training_data/210308/training_data.csv`, the training data file.
+- `vectorizer_types`, e.g. `['count', 'tfidf', 'bert', 'scibert']`, the vectorizers you want to use.
+- `classifier_types`, e.g. `['naive_bayes', 'SVM', 'log_reg']`, the classifier types you want to use.
+
+Then by running 
 ```
-python nutrition_labels/grant_tagger.py --training_data_file data/processed/training_data.csv --vectorizer_type count --relevant_sample_ratio 1 --model_type naive_bayes --bert_type bert
+python nutrition_labels/grant_tagger.py --config_path configs/train_model/2021.03.16.ini
```
Comment (nsorros) [inline README.md:71]: nice way to run models and have them reproducible. in case of interest you can define an entry point like I have done here https://github.com/wellcometrust/grants_tagger/blob/f269800b12ea6081d545a97a573ed44f01c6cb80/setup.py#L26
then you can all `grants_tagger --config_path configs/train_model/2021.03.16.ini`

as far as I know you have multiple entry points or you can also add a CLI like I have done here using Typer https://github.com/wellcometrust/grants_tagger/blob/master/grants_tagger/__main__.py (you can also use argparse) and then introduce commands for train, predict etc so you can do
`grants_tagger train --config_path ...` or `grants_tagger evaluate ...`

File: configs/train_model/2021.02.21.ini [hunk 2]
```diff
@@ -0,0 +1,20 @@
+[DEFAULT]
+version = 2021.02.21
+description = Train all models with 210221 training data.
```
Comment (nsorros) [inline configs/train_model/2021.02.21.ini:3]: generally I think it is better to have config per model, even though I understand that in this case since you are ensembling it possibly makes sense to be able to train multiple models. what would possibly be better in my mind is to have defined a VotingClassifier or WellcomeVotingClassifier that defines the models used and trains those.

I am saying that only because in my mind each config is related to a model and this could be replicated for ensemble models either by training models separately or defining all the models that the ensemble model needs as params of using a custom model or something like VotingClassifier.

This is not a hard rule of course, just a preference.

File: tests/test_grant_tagger.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
+import pytest
+
+import pandas as pd
+
+from nutrition_labels.grant_tagger import GrantTagger
+
+
+training_data = pd.DataFrame(
```
Comment (nsorros) [inline tests/test_grant_tagger.py:8]: not that it makes a difference but you could use a fixture as well here. also a pattern I want to get into lately is that test data is defined in one place and all tests use them

File: nutrition_labels/grant_tagger.py [hunk 4]
```diff
@@ -11,248 +39,358 @@
...
 import os
+import ast
+import json
 
 from nutrition_labels.useful_functions import pretty_confusion_matrix
 
 
-class GrantTagger():
+class GrantTagger:
+    """
```
Comment (nsorros) [inline nutrition_labels/grant_tagger.py:53]: nice docstring

File: nutrition_labels/grant_tagger.py [hunk 5]
```diff
@@ -11,248 +39,358 @@
...
+        if self.vectorizer_type == "count":
             self.vectorizer = CountVectorizer(
-                analyzer='word',
-                token_pattern=r'(?u)\b\w+\b',
-                ngram_range=self.ngram_range
-                )
-        elif self.vectorizer_type == 'tfidf':
+                analyzer="word",
+                token_pattern=r"(?u)\b\w+\b",
+                ngram_range=(1, 2),
```
Comment (nsorros) [inline nutrition_labels/grant_tagger.py:133]: so before that was a param, why is it hard coded now? note btw that you can define either a `params` in the init or a `set_params` method which you can use to pass parameters to the underlying `vectorizer` and `classifier`. all sklearn models implement that interface so you can leverage it and pass in params from the config files that affect the vectorizer and the classifier so you can bette record your experiments.

File: nutrition_labels/grant_tagger.py [hunk 6]
```diff
@@ -11,248 +39,358 @@
...
+        if "bert" in self.vectorizer_type and self.classifier_type == "naive_bayes":
             scaler = MinMaxScaler()
-            X_vect = scaler.fit_transform(X_vect)
+            self.X_vect = scaler.fit_transform(self.X_vect)
 
-        return X_vect, y
+        return self.X_vect, self.y
 
-    def split_data(self, X_vect, y, split_seed):
+    def split_data(self, X_vect, y):
```
Comment (nsorros) [inline nutrition_labels/grant_tagger.py:153]: why split_data is part of the model? usually we split data before passing them to the model. also I notice that there seem to be some data that are deemed irrelevant, what are those? I wonder whether those along with splitting should be dealt in preprocessing before the model ever gets called

File: nutrition_labels/grant_tagger.py [hunk 7]
```diff
@@ -11,248 +39,358 @@
...
+    classifier_types = ast.literal_eval(config["models"]["classifier_types"])
 
-    grant_tagger = GrantTagger(
-        ngram_range=(1, 2),
-        test_size=0.25,
-        vectorizer_type=args.vectorizer_type,
-        model_type=args.model_type,
-        bert_type=args.bert_type,
-        relevant_sample_ratio=args.relevant_sample_ratio
+    relevant_sample_ratio = config.getfloat("params", "relevant_sample_ratio")
```
Comment (nsorros) [inline nutrition_labels/grant_tagger.py:324]: did not know about getfloat, nice

File: nutrition_labels/grant_tagger.py [hunk 8]
```diff
@@ -11,248 +39,358 @@
...
-    outout_name = '_'.join([outout_name, datestamp])
+    config = configparser.ConfigParser()
+    config.read(args.config_path)
 
-    grant_tagger.save_model(
-        os.path.join('models', outout_name),
-        split_info=split_info,
-        evaluation_results=evaluation_results
-        )
+    train_several_models(config)
```
Comment (nsorros) [inline nutrition_labels/grant_tagger.py:396]: also use that minimal pattern of reading the params and calling function 👍

File: nutrition_labels/grant_tagger.py [hunk 9]
```diff
@@ -11,248 +39,358 @@
...
+    config_version = "".join(config["DEFAULT"]["version"].split("."))[2:]
+
+    # Train and save several models
+    for vectorizer_type in vectorizer_types:
+        for classifier_type in classifier_types:
+            print(f"Training for {vectorizer_type} + {classifier_type} ...")
+
+            if not split_seed:
+                split_seed = np.random.randint(100)  # Pick randomly if not given
+            grant_tagger = GrantTagger(
```
Comment (nsorros) [inline nutrition_labels/grant_tagger.py:342]: maybe a space here, purely for visual reasons

File: nutrition_labels/grant_tagger.py [hunk 10]
```diff
@@ -11,248 +39,358 @@
...
+                test_size=test_size,
+                relevant_sample_ratio=relevant_sample_ratio,
+                split_seed=split_seed,
+                vectorizer_type=vectorizer_type,
+                classifier_type=classifier_type,
+                prediction_cols=prediction_cols,
+                label_name=label_name,
+            )
+            X_vect, y = grant_tagger.transform(training_data, train_data_id)
```
Comment (nsorros) [inline nutrition_labels/grant_tagger.py:351]: is that ever needed outside the context of training? cause if not then I would internally define a pipeline with the vectorizer and classifier chosen by the user and only call `fit`, `predict`, `evaluate` and `save`

File: nutrition_labels/grant_tagger.py [hunk 11]
```diff
@@ -11,248 +39,358 @@
...
+
+            X_train, X_test, y_train, y_test = grant_tagger.split_data(X_vect, y)
```
Comment (nsorros) [inline nutrition_labels/grant_tagger.py:353]: as discussed previously I would split data before as a preprocessing step and combined with the comment about the pipeline I would eliminate the need for two calls before the fit. fit would operate on the preprocessed and splitter data.

on a more general note, I think it is worth asking what functionality you are adding on top of sklearn, in other words what prevents you from using sklearn all the way. two additions here seem to be preparing the data by discarding some and splitting them which could be done outside the model or in a separate class called `DataLoader` or something. other than that I see a different evaluate and a save method which justify the use of a custom model. the reason I am mentioning that is that whenever we decide to introduce custom classes we make it slightly less maintainable in the sense that we need to maintain the functionality works not the sklearn community which is fine of course when there is a reason.

**Review by nsorros — APPROVED** [review]
I do not see a reason to block this but I think it is worth thinking whether you want to restructure a bit the model class so that
* preprocessing and splitting of data happens outside
* there is a set_params to pass params in and record any changes in params in the configs

and additionally think whether configs should relate to one model incarceration or multiple. currently you seem to have created a config that recreates a couple of models but are they all used to produce one ensemble? what happens if a user wants to recreate one of them in the future or change one param, do they retrain all of them? on the other hand there might be a good reason for having a config to retrain multiple models as well as configs for one model for example when you get new data as you do, you might want to retrain all of them but I guess the question is why do you want to train all of them instead of only the ones that is the best, the one that is or will be in production.

**Comment by lizgzil** [comment]
> I do not see a reason to block this but I think it is worth thinking whether you want to restructure a bit the model class so that
> 
> * preprocessing and splitting of data happens outside
Yes this is a good idea

> * there is a set_params to pass params in and record any changes in params in the configs
Interesting. So basically create your own config via arguments rather than actually making the .ini file?

> and additionally think whether configs should relate to one model incarceration or multiple. currently you seem to have created a config that recreates a couple of models but are they all used to produce one ensemble? what happens if a user wants to recreate one of them in the future or change one param, do they retrain all of them? on the other hand there might be a good reason for having a config to retrain multiple models as well as configs for one model for example when you get new data as you do, you might want to retrain all of them but I guess the question is why do you want to train all of them instead of only the ones that is the best, the one that is or will be in production.

The config here is just for the training of 12 candidate models, and then after deciding on the best combination for the ensemble. So I suppose at this point in the pipeline it's not known which models are best given new training data/parameters. There is another script and [config file](https://github.com/wellcometrust/nutrition-labels/blob/main/configs/ensemble/2021.01.29.ini) for making predictions using the ensemble model in which you specify which models you want to include. So I suppose in the past I never ran this particular script only once. Having said that, the GrantTagger class is set up to just run one model, and in the config you are able to just specify one vectorizer and one classifier - so if it was ever needed to just run one then it is possible.

**Comment by lizgzil** [comment]
I'm going to merge this and open a new one for something I want to do with the text processing
